cobraproclimits/cobraproclimits.go

```diff
@@ -0,0 +1,63 @@
+package cobraproclimits
+
+import (
+	"log/slog"
+
+	"github.com/KimMachineGun/automemlimit/memlimit"
+	"github.com/jzelinskie/cobrautil/v2"
+	"github.com/rs/zerolog"
+	slogzerolog "github.com/samber/slog-zerolog/v2"
+	"github.com/spf13/cobra"
+	"go.uber.org/automaxprocs/maxprocs"
+)
+
+// NOTE: Both of these assume that there is already a zerolog instance configured for the process
+// by the time this RunE is invoked.
+
+// SetLimitsRunE wraps the RunFunc with setup logic for memory limits
+// for the go process. It requests 90% of the memory available and respects
+// kubernetes cgroup limits.
+func SetMemLimitRunE() cobrautil.CobraRunFunc {
+	return func(cmd *cobra.Command, args []string) error {
+		// Need to invert the slog => zerolog map so that we can get the correct
+		// slog loglevel for memlimit logging
+		logLevelMap := make(map[zerolog.Level]slog.Level, len(slogzerolog.LogLevels))
+		for sLevel, zLevel := range slogzerolog.LogLevels {
+			logLevelMap[zLevel] = sLevel
+		}
+
+		logger := zerolog.DefaultContextLogger
```

Comment:
This makes assumptions on how the logger is wired-up, and does not seem to align well with the zerolog package other than using the same library.

Perhaps an alternative is to make these `*RunE` functions take a logger, or even better, have a fluent-builder API like the other packages.

Reply:
Hmm... yeah, that makes sense on the face of it. I wasn't sure to what degree the stuff in `cobrautil` is meant to be a general-purpose lib vs something that's specific and idiosyncratic to our setup. This PR originally came out of a request from Evan that we move this configuration into a common place that tiger can use as well, and my understanding is that that's a similar environment/setup to spicedb.

I also wanted to defer the cost of making this general-purpose until we had a need for it, but I'm willing to hear arguments for that being now.

Reply:
Yeah I don't think I came up with this out of making this general purpose, but from two points of view:
- `zerolog.DefaultContextLogger` is not a "default global logger" but the default logger used when no logger is present in the `context.Context`. It seems like a confusing default to use with non-obvious outcome, subject to races (depends on the order of something initializing the default context first).
- if a logger argument was added, we would face the same kind of annoying API breakages result of adding new arguments to a function signature. That's one of the main drivers to the `v2` fluent API design, which is the result from us using the library and having to evolve it over time.

I'd need to revisit the logging code to understand what the behaviour would be, but that in itself seems already like a red flag to me that the behaviour is not easy to grok.